### ISSUE TITLE
`cf push` fails with jars generated with the Spring Boot 1.3.0 snapshot release

### ISSUE THREAD

**Issue by dmikusa-pivotal** [cloudfoundry/cli#411]
Steps to reproduce:

  - Create new Maven project w/STS (I added Spring Boot Web Starter, but I don't think it's necessary), pick the 1.3.0 snapshot
  - Run `mvn package`
  - Run `cf push <some-app-name> -p target/demo-0.0.1-SNAPSHOT.jar`.

It will fail with the following:

```
 cf push my-boot-app -p target/demo-0.0.1-SNAPSHOT.jar
Updating app spring-boot-dan in org <org> / space development as <user>...
OK

Uploading <my-boot-app>...
FAILED
Error uploading application.
/var/folders/w4/g5kx6n_16330906jj_6nvqcc0000gp/T/apps009588947 is empty
```

The jar files can be extracted using a tool like `jar` or `unzip` and all of the files are present.

**Comment by dmikusa-pivotal** [comment]
Here's a link to a demo project.  You can use this instead of creating a new project in the repo instructions above.

   https://drive.google.com/file/d/0Byv8UPk_0AMORS0zU1I4OXpsX1k/view?usp=sharing

**Comment by philwebb** [comment]
For a little more background on this issue and it's root cause see https://github.com/spring-projects/spring-boot/issues/1117.

We are trying to use a little known technique with Spring Boot 1.3 of embedding a Bash script at the start of the jar file so that it can be easily run and installed as a service. The [zip file format](http://en.wikipedia.org/wiki/Zip_%28file_format%29) is a little unusual in that a central directory record is usually read from the end of the file, meaning that you can put whatever you want at the front. You can see how we deal with reading file by looking [at this class](https://github.com/spring-projects/spring-boot/blob/3048f44a964ab2390dff26f0818c941ba40bdc56/spring-boot-tools/spring-boot-loader/src/main/java/org/springframework/boot/loader/jar/CentralDirectoryEndRecord.java).

I imagine that your unzip code or whatever library you use might need updating to deal with the additional bytes at the front of the file.

**Comment by philwebb** [comment]
I've updated one of the existing tests to reproduce the issue. See https://github.com/philwebb/cli/commit/a67df0d86fb15ba484ee8edd294d8a30b6e719e5.

Unfortunately it looks like the root cause is coming from [archive/zip in go itself](http://golang.org/src/archive/zip/reader.go) which probably means there isn't going to be an easy fix.

**Comment by philwebb** [comment]
I've raised an issue with `go` and I have [a possible fix](https://github.com/philwebb/go/tree/zip-junk-prefix). Perhaps this patched copy of `archive/zip` could be used by the CLI to work around the problem?

**Comment by goehmen** [comment]
We'll need to watch to see when Go picks up this fix.  We're already looking to upgr to go 1.4.

**Comment by philwebb** [comment]
@goehmen So far they don't seem too keen to merge the fix. I need to find find out what techniques other tools use to deal with this. I'm stacked at the moment so If you have any time to add relevant feedback to https://github.com/golang/go/issues/10464 please do.

**Comment by philwebb** [comment]
@goehmen The suggested patch isn't going to make Go 1.5 but an [alternative fix as been suggested](https://github.com/golang/go/issues/10464#issuecomment-97974790). Can we reopen this one to consider if it's possible to support this ourselves without waiting for Go?

**Comment by goehmen** [comment]
reopened.

Greg Oehmen
Cloud Foundry Product Manager
415.205.6596

**Comment by krishicks** [comment]
@philwebb: When the cli uploads the jar, it unpacks it first. What would the expected behavior be with an embedded file in the jar after unpacking? How would it need to be uploaded?

**Comment by dkoper** [comment]
We're not going to add a dependency on a patched golang library that we then have to maintain ourselves.
Closing this issue as a won't-fix; we'll wait for a fix in golang.

**Comment by philwebb** [comment]
@krishicks The embedded script wouldn't actually need to be uploaded, It's can simply be ignored.
@dkoper You don't necessarily need to patch the golang library. https://github.com/golang/go/issues/10464#issuecomment-97974790 suggests simply scanning forward for the `0x04034b50` header.

**Comment by krishicks** [comment]
Scanning for the headers seems like a fine thing to do, as @bradfitz says.

I'm going to re-open this for now. @dkoper should prioritize as he sees fit.

**Comment by dkoper** [comment]
@philwebb, @krishicks Thanks for the extra clarification. Sounds like there is a clear plan on how to tackle this.

**Comment by jdubois** [comment]
This is a huge issue for us ( https://github.com/jhipster/generator-jhipster/issues/2225 ) as this means no JHipster application can be deployed to CloudFoundry.

**Comment by dkoper** [comment]
@dmikusa-pivotal , @philwebb We have made a fix. Can you try with our edge binary?
https://github.com/cloudfoundry/cli#downloads
Links are in the last paragraph.

Cheers,
Dies Koper
CF CLI PM

**Comment by dmikusa-pivotal** [comment]
Works for me with the edge binary.

**Comment by dkoper** [comment]
Thanks Dan for the quick confirmation!

**Comment by philwebb** [comment]
Sweet. Thanks @dkoper & @simonleung8 for the fix. I realize this is a bit of an edge case so I really appreciate it :+1:

**Comment by krishicks** [comment]
😓

**Comment by dkoper** [comment]
This fix is included in cf CLI release v6.14.1 published today.
https://github.com/cloudfoundry/cli#downloads
